### PR TITLE
Add @superset-ui/time-format

### PR DESCRIPTION
**Pull request by kristw**
🏆 Enhancements

Add `@superset-ui/time-format`, a module for formatting time. Functions `getTimeFormatter` and `formatTime` should be used instead of calling `d3.format` directly.

```js
import { getTimeFormatter } from '@superset-ui/time-format';
const formatter = getTimeFormatter('%Y-%m-d');
console.log(formatter(new Date()));
```

or

```js
import { formatTime } from '@superset-ui/time-format';
console.log(formatTime('%Y-%m-d', new Date()));
```

It is powered by a registry to support registration of custom formatting, with fallback to `d3.utcFormat` or `d3.timeFormat` (if the formatId starts with `local!`)

```js
import { getTimeFormatterRegistry, formatTime, TimeFormatter } from '@superset-ui/time-format';

getTimeFormatterRegistry().registerValue('my_format', new TimeFormatter({
  id: 'my_format',
  formatFunc: v => `my special format of ${utcFormat('%Y')(v)}`
});

console.log(formatTime('my_format', new Date(2018)));
// prints 'my special format of 2018'
```

It also define constants for common d3 time formats. See `TimeFormats.js`.

```js
import { TimeFormats } from '@superset-ui/time-format';

TimeFormats.DATABASE_DATETIME // '%Y-%m-%d %H:%M:%S'
TimeFormats.US_DATE // '%m/%d/%Y'
```

### CODE REVIEW

**Comment by kristw** [comment]
Just made updates and this one should be ready to review.

File: packages/superset-ui-time-format/README.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+## @superset-ui/time-format
+
+[![Version](https://img.shields.io/npm/v/@superset-ui/time-format.svg?style=flat)](https://img.shields.io/npm/v/@superset-ui/time-format.svg?style=flat)
+[![David (path)](https://img.shields.io/david/apache-superset/superset-ui.svg?path=packages%2Fsuperset-ui-time-format&style=flat-square)](https://david-dm.org/apache-superset/superset-ui?path=packages/superset-ui-time-format)
+
+Description
+
+#### Example usage
```
[the remark this answers is not in the document]
Comment (kristw) [reply packages/superset-ui-time-format/README.md]: 👍

File: packages/superset-ui-time-format/test/TimeFormatter.test.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      expect(formatter.format(PREVIEW_TIME)).toEqual('2017');
+    });
+  });
+  describe('.preview(value)', () => {
+    const formatter = new TimeFormatter({
+      id: 'year_only',
+      formatFunc: value => `${value.getFullYear()}`,
+    });
+    it('returns string comparing value before and after formatting', () => {
+      expect(formatter.preview(PREVIEW_TIME)).toEqual('Tue, 14 Feb 2017 11:22:33 GMT => 2017');
```
[the remark this answers is not in the document]
Comment (kristw) [reply packages/superset-ui-time-format/test/TimeFormatter.test.js]: Ah true. Will change day.

File: packages/superset-ui-time-format/src/TimeFormatter.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import { ExtensibleFunction, isRequired } from '@superset-ui/core';
+
+// eslint-disable-next-line
```
[the remark this answers is not in the document]
Comment (kristw) [reply packages/superset-ui-time-format/src/TimeFormatter.js]: `no-magic-numbers`

File: packages/superset-ui-time-format/src/TimeFormatterRegistry.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+  get(format) {
+    const targetFormat = format || this.defaultKey;
+
+    if (this.has(targetFormat)) {
+      return super.get(targetFormat);
+    }
+
+    // Create new formatter if does not exist
+    const useLocalTime = targetFormat.startsWith(LOCAL_PREFIX);
```
[the remark this answers is not in the document]
Comment (williaster) [reply packages/superset-ui-time-format/src/TimeFormatterRegistry.js]: ah I didn't realize this complexity, makes sense 👍

**Pull request merged**